### PR TITLE
Added toggle function

### PR DESCRIPTION
**Pull request by babusid**

### CODE REVIEW

**Review by IshDeshpa — APPROVED** [review]

**Review by Cam0Cow — APPROVED** [review]

File: Apps/Src/BlinkLights.c [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#include "Tasks.h"
+#include "Lights.h"
+
+/**
+* @brief This thread controls the blinking of the hazard and indicator lights at the regulation specified frequency. 
+**/
+
+void Task_BlinkLight(void* p_arg){
```
Comment (ClarkPoon) [inline Apps/Src/BlinkLights.c:8]: Why did you add this?

File: Drivers/Src/Lights.c [hunk 2]
```diff
@@ -84,6 +84,34 @@ uint16_t Lights_Bitmap_Read(light_t light) {
...
+ * @brief Toggles a light. Should be used only after Toggle_Enable has been called for this light so that we are accurately tracking the enabled and disabled lights
+ * @param light Which light to toggle
+*/
+void Toggle_Light(light_t light){
+    State lightState = Lights_Read(light);
+    if(lightState == OFF){
+        Lights_Set(light, ON);
+    } else {
+        Lights_Set(light, OFF);
+    }
```
Comment (ClarkPoon) [inline Drivers/Src/Lights.c:112]: This is not thread-safe, since you can get interrupted in between reading and setting the light's state. Not sure if you are relying on this to be thread-safe or not

**Review by tiandahuang — APPROVED** [review]

File: Apps/Src/BlinkLights.c (hunk 1 above)
Comment (IshDeshpa) [reply Apps/Src/BlinkLights.c:8]: oops that was meant to be in the new blinklights issue i'm working on

File: Drivers/Src/Lights.c (hunk 2 above)
Comment (IshDeshpa) [reply Drivers/Src/Lights.c:112]: The only task calling this should be BlinkLights, so I think it should be OK

File: Drivers/Src/Lights.c [hunk 3]
```diff
@@ -84,6 +84,41 @@ uint16_t Lights_Bitmap_Read(light_t light) {
...
+}
+
+/**
+ * @brief Toggles multiple lights according to the toggle bitmap
+*/
+void Lights_MultiToggle(){
```
Comment (Cam0Cow) [inline Drivers/Src/Lights.c:103]: Same change as in header file

File: Drivers/Src/Lights.c [hunk 4]
```diff
@@ -99,6 +134,7 @@ State Toggle_Read(light_t light) {
  */
 uint16_t Toggle_Bitmap_Read() {
```
Comment (Cam0Cow) [inline Drivers/Src/Lights.c:135]: Same change as header file

File: Drivers/Src/Lights.c [hunk 5]
```diff
@@ -84,6 +84,41 @@ uint16_t Lights_Bitmap_Read(light_t light) {
...
+ * @param   light Which light to enable toggling for
+ * @param   state State to set toggling
+ * @return  void
+ */
+void Toggle_Set(light_t light, State state) {
+    // Mutex not needed here because only BlinkLights uses this bitmap
+    if(state==ON)
+        lightToggleBitmap |= (0x01<<light);
+    else if(state==OFF)
+        lightToggleBitmap &= ~(0x01<<light);
```
Comment (Cam0Cow) [inline Drivers/Src/Lights.c:119]: This hurts my eyes, please use curly braces around the blocks.

**Review by Cam0Cow — CHANGES_REQUESTED** [review]

**Review by Cam0Cow — APPROVED** [review]

**Review by Cam0Cow — DISMISSED** [review]

File: Drivers/Inc/Lights.h [hunk 6]
```diff
@@ -50,5 +50,63 @@ State Lights_Read(light_t light);
...
+
+ * @brief   Set light toggling
+ * @param   light Which light to enable toggling for
+ * @param   state State to set toggling
+ * @return  void
+ */
+void Toggle_Set(light_t light, State state);
+
+/**
+* @brief   Read the state of a specific toggleable light from the toggle bitmap
```
Comment (babusid) [inline Drivers/Inc/Lights.h:82]:
```suggestion
* @brief   Read the toggle enable state of a specific light from the toggle bitmap
```

File: Drivers/Inc/Lights.h [hunk 7]
```diff
@@ -50,5 +50,63 @@ State Lights_Read(light_t light);
...
+uint16_t Toggle_Bitmap_Read(void);
+
+
+/**
+ * @brief   Set light to given state
+ * @param   light Which light to set
+ * @param   state What state to set the light to
+ * @return  void
+ */
+void Lights_Set(light_t light, State state);
```
Comment (babusid) [inline Drivers/Inc/Lights.h:102]: This is already at the top of the header file
```suggestion
```

**Review by IshDeshpa — APPROVED** [review]

**Review by Cam0Cow — APPROVED** [review]
